README
================================================================================
=                 Bluefog - Generate Phantom Bluetooth Devices                 =
=                       Tom Nardi (MS3FGX@gmail.com)                           =
================================================================================

Bluefog is able to generate an essentially unlimited number of phantom
Bluetooth devices, useful for testing Bluetooth scanning and monitoring
systems, making it more difficult for attackers to lock onto your devices, or
otherwise complicate the normal operation of Bluetooth devices.

Technically Bluefog can work with just one Bluetooth adapter, but it works much
better when you connect up multiple adapters. Currently Bluefog can support 6
simultaneous Bluetooth radios, but that value may go up as I experiment with
it some more.

--------------------------------------------------------------------------------
- WARNINGS                                                                     -
--------------------------------------------------------------------------------

The changes Bluefog makes to your Bluetooth hardware is supposed to be
temporary, but not all Bluetooth chipsets behave appropriately. During the
development of Bluefog, I managed to permanently delete the MAC address of
my laptop's internal radio. This isn't that big of a deal, but you may want to
keep it in mind.

In addition, I have occasionally found BlueZ to lock up after running Bluefog
with large numbers of Bluetooth radios (4 or more), and indeed, BlueZ in
general doesn't seem to work reliably with multiple radios. If BlueZ starts to
behave oddly after running Bluefog, just reboot and everything should be back
to normal.

--------------------------------------------------------------------------------
- Compatibility                                                                -
--------------------------------------------------------------------------------

To date, Bluefog has only been tested with Bluetooth adapters based on chipsets
from Cambridge Silicon Radio (CSR) and Broadcom. Compatibility reports in
regards to other hardware would be greatly appreciated.

--------------------------------------------------------------------------------
- Installation                                                                 -
--------------------------------------------------------------------------------

To compile Bluefog, simply run the command "make" in this directory. If you get
any message about missing libraries, use your distribution's package manager to
install them. This will let you run Bluefog from the current directory for a
quick test, but not install it to the system.

To actually install Bluefog, use the command "make install". Since this will
install Bluefog to the system directories, you will need to run this as root or
through sudo.

When and if you want to remove Bluefog, you would use the command "make
uninstall". As with "install", this makes changes to the system directories and
will therefore require root permissions.

In addition, if you are upgrading from a previous version, there is also the
command "make upgrade". This will remove any older versions of Bluefog, compile
the new version, and finally install it to the system. As you might have
guessed, this also requires root permissions.

--------------------------------------------------------------------------------
- Usage                                                                        -
--------------------------------------------------------------------------------

As of the current version, Bluefog supports the following options:

-i <interface>
    Allows you to manually select which interface Bluefog uses. Without this
option, Bluefog will automatically detect the first Bluetooth device and use
that.

If you've chosen to use multiple threads, Bluefog will use all devices
available, so this option won't have any effect.

-t <threads>
    With this option you can set how many parallel jobs you want Bluefog to
perform. This number should match however many physical Bluetooth radios your
computer currently has.

The maximum number of threads is currently 6 (which obviously requires 6
separate Bluetooth adapters to be on the system), but may change in the future.

-n <name>
    Usually, Bluefog will chose a random name from the hundreds of names
included with the software. If you use this option however, you can set Bluefog
to use the same custom name for all devices it spoofs. This option is great at
parties.

-d <seconds>
    This determines how many seconds to wait between device spoofs. The minimum
value for this option is 5 seconds, but realistically, anything under 20
probably won't work very well. Default is 30, which is rather slow but at least
has a good shot at working.

-l
    This enables "Loiter Mode", which will randomly slow down the spoofing of
devices. This is used to simulate slow moving targets, rather than a constant
stream. This mode will create a more realistic flow of phantom devices, closer
to what you would expect during real Bluetooth monitoring.

-m
    This option allows you to toggle the randomization of Bluetooth MAC
addresses by Bluefog. Realistically, there isn't much point in randomizing
the device name if you don't change the MAC (as most devices ignore the name
anyway), but I've included the option anyway.

The default is ENABLED, so most people can safely ignore this.

-c
    This option allows you to toggle the randomization of the spoofed device's
class information. If you leave this option off, Bluefog will set all spoofed
device's class to that of a standard smartphone.

-v
    This option allows you to toggle verbose mode. With verbose mode on, you'll
be able to see the hardware being initialized, and see a message in the
terminal every time a new device is spoofed.

--------------------------------------------------------------------------------
- Acknowledgements                                                             -
--------------------------------------------------------------------------------

Bluefog was greatly inspired by "SpoofTooph", written by JP Dunning (.ronin).
In addition to the spiritual support, Bluefog takes a few functions from
SpoofTooph in the areas of class and MAC address randomization, and includes
the data collected by ronin's "Bluetooth Profiling Project".

You can read more about JP Dunning's projects on his site:
http://www.hackfromacave.com/

Bluefog uses quite a bit of code from the "BlueZ" project by Marcel Holtmann,
most notably the majority of bdaddr.c, plus some functions from hciconfig.c.

For more information, check out the official BlueZ site:
http://www.bluez.org/

Bluefog also contains data collected by Brad Haines (RenderMan) during his
recent scans of the West Edmonton Mall in Alberta, Canada.

You can read more about RenderMan's projects on his site:
http://www.renderlab.net/

--------------------------------------------------------------------------------
- License                                                                      -
--------------------------------------------------------------------------------

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License version 2 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

For details, see the file "COPYING" in the source directory.

--------------------------------------------------------------------------------
- More Info                                                                    -
--------------------------------------------------------------------------------

For more information and updates, please see www.digifail.com


ChangeLog
03/31/13:
	devicenames.h touch up

03/26/13:
	Removed unused oui.c code

02/09/13: Released as Version 0.0.4
	Improved generation of random MAC
	Updated README

01/26/13:
	Only verify changed MAC on first spoof
	Added Loiter mode
	Updated README

01/13/13: Released as Version 0.0.3
	Fixed mistake I made in merging device name files

01/11/12:
	Implemented baby's first sigaction, threads sorta cleanly exit
	Cleanups throughout

12/31/12:
	Improvements to device name loading
	Fix buffer overflow in device naming
	Name list fixes

12/28/12:
	Minor fixes
	Added basic groundwork for signal handler, needs work

12/23/12:
	Raised maximum radio count to 6

12/21/12: End Of The World Edition
	MAC spoofing now works on Broadcom and CSR
	Set default class when not randomized
	Added -n option to specify static (troll) device name
	Improve status info on startup
	Integrated data from "Operation Street Sweep"
	Rough -h output added

12/19/12:
	Code for detecting if interface MAC is changing
	Added -c for randomizing class
	Removed iteration parameter, fog now runs indefinitely 

12/18/12:
	Development begins


bdaddr.c
/*
 *  Originally from BlueZ, by Marcel Holtmann <marcel@holtmann.org>
 *
 *  Modified by JP Dunning (.ronin) for SpoofTooph
 *
 *  Additional modifications by Tom Nardi (MS3FGX) for Bluefog
 * 
 *  Released under the GPLv2, see "COPYING".
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <getopt.h>
#include <sys/ioctl.h>
#include <sys/socket.h>

#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>

static int transient = 1;

static int generic_reset_device(int dd)
{
	bdaddr_t bdaddr;
	int err;

	err = hci_send_cmd(dd, 0x03, 0x0003, 0, NULL);
	if (err < 0)
		return err;

	return hci_read_bd_addr(dd, &bdaddr, 10000);
}

#define OCF_ERICSSON_WRITE_BD_ADDR	0x000d
typedef struct {
	bdaddr_t	bdaddr;
} __attribute__ ((packed)) ericsson_write_bd_addr_cp;
#define ERICSSON_WRITE_BD_ADDR_CP_SIZE 6

static int ericsson_write_bd_addr(int dd, bdaddr_t *bdaddr)
{
	struct hci_request rq;
	ericsson_write_bd_addr_cp cp;

	memset(&cp, 0, sizeof(cp));
	bacpy(&cp.bdaddr, bdaddr);

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = OCF_ERICSSON_WRITE_BD_ADDR;
	rq.cparam = &cp;
	rq.clen   = ERICSSON_WRITE_BD_ADDR_CP_SIZE;
	rq.rparam = NULL;
	rq.rlen   = 0;

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	return 0;
}

#define OCF_ERICSSON_STORE_IN_FLASH	0x0022
typedef struct {
	uint8_t		user_id;
	uint8_t		flash_length;
	uint8_t		flash_data[253];
} __attribute__ ((packed)) ericsson_store_in_flash_cp;
#define ERICSSON_STORE_IN_FLASH_CP_SIZE 255

static int ericsson_store_in_flash(int dd, uint8_t user_id, uint8_t flash_length, uint8_t *flash_data)
{
	struct hci_request rq;
	ericsson_store_in_flash_cp cp;

	memset(&cp, 0, sizeof(cp));
	cp.user_id = user_id;
	cp.flash_length = flash_length;
	if (flash_length > 0)
		memcpy(cp.flash_data, flash_data, flash_length);

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = OCF_ERICSSON_STORE_IN_FLASH;
	rq.cparam = &cp;
	rq.clen   = ERICSSON_STORE_IN_FLASH_CP_SIZE;
	rq.rparam = NULL;
	rq.rlen   = 0;

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	return 0;
}

static int csr_write_bd_addr(int dd, bdaddr_t *bdaddr)
{
	unsigned char cmd[] = { 0x02, 0x00, 0x0c, 0x00, 0x11, 0x47, 0x03, 0x70,
				0x00, 0x00, 0x01, 0x00, 0x04, 0x00, 0x00, 0x00,
				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };

	unsigned char cp[254], rp[254];
	struct hci_request rq;

	if (transient)
		cmd[14] = 0x08;

	cmd[16] = bdaddr->b[2];
	cmd[17] = 0x00;
	cmd[18] = bdaddr->b[0];
	cmd[19] = bdaddr->b[1];
	cmd[20] = bdaddr->b[3];
	cmd[21] = 0x00;
	cmd[22] = bdaddr->b[4];
	cmd[23] = bdaddr->b[5];

	memset(&cp, 0, sizeof(cp));
	cp[0] = 0xc2;
	memcpy(cp + 1, cmd, sizeof(cmd));

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = 0x00;
	rq.event  = EVT_VENDOR;
	rq.cparam = cp;
	rq.clen   = sizeof(cmd) + 1;
	rq.rparam = rp;
	rq.rlen   = sizeof(rp);

	if (hci_send_req(dd, &rq, 2000) < 0)
		return -1;

	if (rp[0] != 0xc2) {
		errno = EIO;
		return -1;
	}

	if ((rp[9] + (rp[10] << 8)) != 0) {
		errno = ENXIO;
		return -1;
	}

	return 0;
}

static int csr_reset_device(int dd)
{
	unsigned char cmd[] = { 0x02, 0x00, 0x09, 0x00,
				0x00, 0x00, 0x01, 0x40, 0x00, 0x00,
				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };

	unsigned char cp[254], rp[254];
	struct hci_request rq;
	
	if (transient)
		cmd[6] = 0x02;

	memset(&cp, 0, sizeof(cp));
	cp[0] = 0xc2;
	memcpy(cp + 1, cmd, sizeof(cmd));

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = 0x00;
	rq.event  = EVT_VENDOR;
	rq.cparam = cp;
	rq.clen   = sizeof(cmd) + 1;
	rq.rparam = rp;
	rq.rlen   = sizeof(rp);
	
	if (hci_send_req(dd, &rq, 2000) < 0)
		return -1;
	return 0;
}

#define OCF_TI_WRITE_BD_ADDR		0x0006
typedef struct {
	bdaddr_t	bdaddr;
} __attribute__ ((packed)) ti_write_bd_addr_cp;
#define TI_WRITE_BD_ADDR_CP_SIZE 6

static int ti_write_bd_addr(int dd, bdaddr_t *bdaddr)
{
	struct hci_request rq;
	ti_write_bd_addr_cp cp;

	memset(&cp, 0, sizeof(cp));
	bacpy(&cp.bdaddr, bdaddr);

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = OCF_TI_WRITE_BD_ADDR;
	rq.cparam = &cp;
	rq.clen   = TI_WRITE_BD_ADDR_CP_SIZE;
	rq.rparam = NULL;
	rq.rlen   = 0;

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	return 0;
}

#define OCF_BCM_WRITE_BD_ADDR		0x0001
typedef struct {
	bdaddr_t	bdaddr;
} __attribute__ ((packed)) bcm_write_bd_addr_cp;
#define BCM_WRITE_BD_ADDR_CP_SIZE 6

static int bcm_write_bd_addr(int dd, bdaddr_t *bdaddr)
{
	struct hci_request rq;
	bcm_write_bd_addr_cp cp;

	memset(&cp, 0, sizeof(cp));
	bacpy(&cp.bdaddr, bdaddr);

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = OCF_BCM_WRITE_BD_ADDR;
	rq.cparam = &cp;
	rq.clen   = BCM_WRITE_BD_ADDR_CP_SIZE;
	rq.rparam = NULL;
	rq.rlen   = 0;

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	return 0;
}

#define OCF_ZEEVO_WRITE_BD_ADDR		0x0001
typedef struct {
	bdaddr_t	bdaddr;
} __attribute__ ((packed)) zeevo_write_bd_addr_cp;
#define ZEEVO_WRITE_BD_ADDR_CP_SIZE 6

static int zeevo_write_bd_addr(int dd, bdaddr_t *bdaddr)
{
	struct hci_request rq;
	zeevo_write_bd_addr_cp cp;

	memset(&cp, 0, sizeof(cp));
	bacpy(&cp.bdaddr, bdaddr);

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = OCF_ZEEVO_WRITE_BD_ADDR;
	rq.cparam = &cp;
	rq.clen   = ZEEVO_WRITE_BD_ADDR_CP_SIZE;
	rq.rparam = NULL;
	rq.rlen   = 0;

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	return 0;
}

static int st_write_bd_addr(int dd, bdaddr_t *bdaddr)
{
	return ericsson_store_in_flash(dd, 0xfe, 6, (uint8_t *) bdaddr);
}

static struct {
	uint16_t compid;
	int (*write_bd_addr)(int dd, bdaddr_t *bdaddr);
	int (*reset_device)(int dd);
} vendor[] = {
	{ 0,		ericsson_write_bd_addr,	NULL			},
	{ 10,		csr_write_bd_addr,	csr_reset_device	},
	{ 13,		ti_write_bd_addr,	NULL			},
	{ 15,		bcm_write_bd_addr,	generic_reset_device	},
	{ 18,		zeevo_write_bd_addr,	NULL			},
	{ 48,		st_write_bd_addr,	generic_reset_device	},
	{ 57,		ericsson_write_bd_addr,	generic_reset_device	},
	{ 65535,	NULL,			NULL			},
};

static int cmd_bdaddr(int dev, int dd, char * new_addr)
{
  struct hci_dev_info di;
	struct hci_version ver;
	bdaddr_t bdaddr;
	char addr[18], oui[9];
	int i;

	bacpy(&bdaddr, BDADDR_ANY);

	optind = 0;
	
	if (hci_devinfo(dev, &di) < 0) {
		fprintf(stderr, "Can't get device info for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (hci_read_local_version(dd, &ver, 1000) < 0) {
		fprintf(stderr, "Can't read version info for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (!bacmp(&di.bdaddr, BDADDR_ANY)) {
		if (hci_read_bd_addr(dd, &bdaddr, 1000) < 0) {
			fprintf(stderr, "Can't read address for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
			hci_close_dev(dd);
			exit(1);
		}
	} else
		bacpy(&bdaddr, &di.bdaddr);

	str2ba(new_addr, &bdaddr);

	if (!bacmp(&bdaddr, BDADDR_ANY)) {
		printf("Invalid MAC! Something has gone horribly wrong!\n");
		hci_close_dev(dd);
		return(0);
	}
	
	for (i = 0; vendor[i].compid != 65535; i++)
		if (ver.manufacturer == vendor[i].compid) {
			ba2oui(&bdaddr, oui);
			ba2str(&bdaddr, addr);
			
			if (vendor[i].write_bd_addr(dd, &bdaddr) < 0) {
				fprintf(stderr, "Can't write new address\n");
				hci_close_dev(dd);
				return(1);
			}
		
			if (vendor[i].reset_device(dd) < 0)
				return(2);
			else
				ioctl(dd, HCIDEVRESET, dev);

			// Everything worked
			return(0);
		}

	hci_close_dev(dd);

	printf("\n");
	fprintf(stderr, "Unsupported manufacturer\n");

	return(1);
};


bluefog.c
/* 
 *  Bluefog - Create phantom Bluetooth devices
 * 
 *  Bluefog is a tool to create phantom Bluetooth devices. Can be used
 *  to confuse attackers or test Bluetooth detection systems.
 * 
 *  Written by Tom Nardi (MS3FGX@gmail.com), released under the GPLv2.
 *  For more information, see: www.digifail.com
 */

#include <time.h>
#include <stdio.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <pthread.h>
#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>

// List of device names
#include "devicenames.h"

// MAC changing functions
#include "bdaddr.c"

#define VERSION	"0.0.4"
#define APPNAME "Bluefog"

// Sane defaults
#define MAX_THREADS 6
#define THREAD_DELAY 1
#define MAX_DELAY 300
#define SHUTDOWN_WAIT 2
#define MAX_LOITER 10
#define LOITER_CHANCE 2
#define DEFAULT_CLASS "0x5a020c"

// Global variables
int verbose = 0;
int end_threads = 0;
int c_count = 0;

// Data to pass to threads
struct thread_data
{
	int thread_id;
	int change_addr;
	int change_class;
	char *static_name;
	int device;
	int delay;
	int loiter;
};

struct thread_data thread_data_array[MAX_THREADS];

void sig_catch(int sig)
{
	// Do something here someday
}

char* get_localtime()
{
	// Time variables
	time_t rawtime;
	struct tm * timeinfo;
	static char time_string[20];
	
	// Find time and put it into time_string
	time (&rawtime);
	timeinfo = localtime(&rawtime);
	strftime(time_string,20,"%D %T",timeinfo);
	
	// Send it back
	return(time_string);
}

// Open BT socket and return ID
int get_bt_socket (int device)
{
	int bt_socket;	
	bt_socket = hci_open_dev(device);
	if (bt_socket < 0)
	{
		printf("Failed to initalize hci%i!\n", device);
		exit(1);
	}
	return(bt_socket);
}

// Write class info
int write_class (int bt_socket, int device, char *class)
{
	uint32_t cod = strtoul(class, NULL, 16);
	if (hci_write_class_of_dev(bt_socket, cod, 2000) < 0)
		fprintf(stderr,"Can't write local class of device on hci%d: %s (%d)\n", device, strerror(errno), errno);

	return(0);
}

// Select random name from list within range
char* random_name (void)
{	
	return (device_name[(rand() % DEV_NAMES)]);	
}

// Generate random MAC address
// Adapted from "SpoofTooph" by JP Dunning (.ronin)
char* random_addr (void)
{	
	char addr_part[3] = {0};
	static char addr[18] = {0};
	int i = 0;
	
	// Fill in the middle
	while ( i < 14)
	{
		sprintf(addr_part, "%02x", (rand() % 254));	
		addr[i++] = addr_part[0];
		addr[i++] = addr_part[1];
		addr[i++] = ':';
	}

	// Tack 2 more random characters to finish it
	sprintf(addr_part, "%02x", (rand() % 254));	
	addr[i++] = addr_part[0];
	addr[i++] = addr_part[1];
		
	return(addr);
}

void *thread_spoof(void *threadarg)
{			
	// Define variables from struct
	int thread_id, change_addr, device, delay, loiter, change_class;
	char *static_name;
	
	// Local use variables
	int bt_socket;
	int first_run = 1;
	int new_dev = 1;
	int loiter_time = 0;

	// Pull data out of struct
	struct thread_data *local_data;
	local_data = (struct thread_data *) threadarg;
	thread_id = local_data->thread_id;
	device = local_data->device;
	change_addr = local_data->change_addr;
	delay = local_data->delay;
	change_class = local_data->change_class;
	static_name = local_data->static_name;
	loiter = local_data->loiter;
	
	// MAC struct
	bdaddr_t bdaddr;
	bacpy(&bdaddr, BDADDR_ANY);
	struct hci_dev_info di;
	char addr_ref[18] = {0};
	char addr_buff[18] = {0};
	
	// Buffers for verbose, ridiculous
	char *addr_buffer = {0};
	char *name_buffer = {0};
	
	// For discoverability
	struct hci_dev_req dr;
	dr.dev_id  = device;
	dr.dev_opt = SCAN_PAGE | SCAN_INQUIRY;
	
	// Device class
	char class[9] = {0};
		
	// Init device	
	if (verbose)
		printf("Initalizing hci%i on thread %i.\n", device, thread_id);

	// Get a socket
	bt_socket = get_bt_socket(device);
		
	// Get MAC for reference
	if (hci_read_bd_addr(bt_socket, &bdaddr, 1000) < 0)
	{
			fprintf(stderr, "Can't read address for hci%d: %s (%d)\n", device, strerror(errno), errno);
			exit(1);
	}
	
	// Original MAC stored to addr_ref
	ba2str(&bdaddr, addr_ref);
	
	while (end_threads != 1)
	{				
		// Do we want to spoof a new device?
		if (new_dev)
		{
			// Attempt to change address first, since it probably requires reset
			if (change_addr)
			{
				addr_buffer = random_addr();
				if (cmd_bdaddr(device, bt_socket, addr_buffer) == 2)
				{
					// This type of device needs to be manually restarted
					hci_close_dev(bt_socket);
					sleep(SHUTDOWN_WAIT);
					bt_socket = get_bt_socket(device);
				}
			}
			
			// Always change name
			if (static_name != NULL)
				name_buffer = static_name;
			else
				name_buffer = random_name();
			
			if (hci_write_local_name(bt_socket, name_buffer, 2000) < 0)
				fprintf(stderr, "Can't change local name on hci%d: %s (%d)\n", device, strerror(errno), errno);
				
			// Change class
			if (change_class)
			{
				// Generate a random class
				int major = (rand() % 9);
				int minor = (rand() % 6);
		
				// Randomly add service flags
				int flag = 0;              
				if (rand() % 2) flag += 1;
				if (rand() % 2) flag += 2;
				if (rand() % 2) flag += 4;
				if (rand() % 2) flag += 8;
				if (rand() % 2) flag += 10;
				if (rand() % 2) flag += 20;
				if (rand() % 2) flag += 40;
				if (rand() % 2) flag += 80;

				// Put class into string
				sprintf(class, "0x%02x%02x%02x", flag, major, minor);		
				write_class(bt_socket, device, class);
			}
			else
			{
				strcpy(class, DEFAULT_CLASS);		
			}	write_class(bt_socket, device, class);

			// Make discoverable
			if (ioctl(bt_socket, HCISETSCAN, (unsigned long) &dr) < 0)
			{
				fprintf(stderr, "Can't set scan mode on hci%d: %s (%d)\n", device, strerror(errno), errno);
				exit(1);
			}

			// Print device if verbose
			if (verbose)
				printf("hci%i - Name: %s Class: %s MAC: %s\n", device, name_buffer, class, addr_buffer);
		}
		
		// Wait
		sleep(delay);
							
		// Only check this once, to save time later
		if (first_run)
		{
			// Verify MAC actually changed
			if (!bacmp(&di.bdaddr, BDADDR_ANY))
			{
				if (hci_read_bd_addr(bt_socket, &bdaddr, 1000) < 0)
				{
					fprintf(stderr, "Can't read address for hci%d: %s (%d)\n", device, strerror(errno), errno);
					exit(1);
				}
			}
			else
				bacpy(&bdaddr, &di.bdaddr);	

			// Test MAC to addr_buff
			ba2str(&bdaddr, addr_buff);	
				
			if ((strcmp (addr_ref, addr_buff) == 0))
			{
				printf("MAC on interface hci%i is not changing. Hardware is likely not compatible.\n", device);
				printf("Disabling MAC changing for this interface. See README for more info.\n");
				change_addr = 0;
			}
			
			// Don't do this again
			first_run = 0;
		}
		
		// Determine if we will loiter a bit
		if (loiter)
		{
			if (loiter_time == 0)
			{
				// Not currently loitering, should we?
				if ((rand() % MAX_LOITER) > (MAX_LOITER / LOITER_CHANCE))
				{
					// How long to wait
					loiter_time = (rand() % MAX_LOITER);
					new_dev = 0;
				}
				else
					new_dev = 1;
			}
			else if (loiter_time >= 1)
				loiter_time--;
		}		
	}
	
	// Close device
	hci_close_dev(bt_socket);

	if (verbose)
		printf("Thread %i done.\n", thread_id);
	
	// Leap home
	pthread_exit(NULL);
}

static void help(void)
{
	printf("%s (v%s) by Tom Nardi \"MS3FGX\" (MS3FGX@gmail.com)\n", APPNAME, VERSION);
	printf("----------------------------------------------------------------\n");
	printf("Bluefog is a tool used to create phantom Bluetooth devices. It can\n"
		"be used to confuse attackers or test Bluetooth detection systems.\n");
	printf("\n");
	printf("At the moment, Bluefog has only been tested with CSR and Broadcom\n"
			"chipsets. Compatibility reports from other devices would be\n"
			"greatly appreciated\n");
	printf("\n");
	printf("For more information, see www.digifail.com\n");
	printf("\n");
	printf("Options:\n"
		"\t-i <interface>   Manually select device to use, default is automatic\n"
		"\t-t <threads>     Set this to match how many Bluetooth adapters you have\n"
		"\t-n <name>        Sets a static device name instead of random\n"
		"\t-d <seconds>     How many seconds to wait between spoofs, default 30\n"
		"\t-l               Loiter, simulates slow moving or stationary devices\n"
		"\t-m               Toggle randomization of MAC address, default is enabled\n"
		"\t-c               Toggle randomization of class info, default disabled\n"
		"\t-v               Toggle verbose messages, default disabled\n"
		"\n");
}

static struct option main_options[] = {
	{ "interface", 1, 0, 'i' },
	{ "threads", 1, 0, 't' },
	{ "name", 1, 0, 'n' },
	{ "delay", 1, 0, 'd' },
	{ "class", 1, 0, 'c' },
	{ "loiter", 0, 0, 'l' },	
	{ "mac", 0, 0, 'm' },
	{ "verbose", 0, 0, 'v' },
	{ "help", 0, 0, 'h' },
	{ 0, 0, 0, 0 }
};
 
int main(int argc, char *argv[])
{		
	// General variables
	int t, opt;
	
	// Thread ID
	pthread_t threads[MAX_THREADS];
		
	// MAC struct
	bdaddr_t bdaddr;
	bacpy(&bdaddr, BDADDR_ANY);
	char addr[19] = {0};
	
	// Options
	int numthreads = 1;
	int delay = 30;
	int change_addr = 1;
	int change_class = 1;
	int device = -1;
	int loiter = 0;	
	char *static_name = NULL;

	// Process options
	while ((opt=getopt_long(argc, argv, "+t:d:i:n:mchlv", main_options, NULL)) != EOF)
	{
		switch (opt)
		{
		case 'i':
			if (!strncasecmp(optarg, "hci", 3))
				hci_devba(atoi(optarg + 3), &bdaddr);
			else
				str2ba(optarg, &bdaddr);
			break;		
		case 't':
			numthreads = atoi(optarg);		
			if (numthreads > MAX_THREADS || numthreads <= 0)
			{
				printf("Invalid number of threads. See README.\n");
				exit(1);
			}
			break;
		case 'd':
			delay = atoi(optarg);		
			if (delay > MAX_DELAY || delay <= 4)
			{
				printf("Invalid delay value. See README.\n");
				exit(1);
			}
			break;			
		case 'l':
			loiter = 1;
		break;
		case 'c':
			change_class = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'm':
			change_addr = 0;
			break;
		case 'n':
			static_name = strdup(optarg);			
			break;
		case 'h':
			help();
			exit(0);
		default:
			printf("Unknown option. Use -h for help, or see README.\n");
			exit(0);
		}
	}
	
	// Check if we are running as root
	if(getuid() != 0)
	{
		printf("You need to be root to run Bluefog!\n");
		exit(1);
	}
			
	// Seed PRNG
	srand(time(NULL));
	
	// Boilerplate
	printf("%s (v%s) by MS3FGX\n", APPNAME, VERSION);
	printf("---------------------------\n");
				
	// Select hardware
	printf("Bluetooth Interface: ");
	ba2str(&bdaddr, addr);
	if (!strcmp(addr, "00:00:00:00:00:00"))
	{
		printf("Auto\n");
	}
	else
	{
		numthreads = 1;
		device = hci_devid(addr);		
		printf("hci%i\n", device);
	}

	// Static or random names
	printf("Device Name: ");
	if (static_name == NULL)
	{
		printf("Randomized\n");
		
		// Number of names loaded
		printf("Available Names: %i\n", DEV_NAMES - 1);
	}
	else
		printf("Static (%s)\n", static_name);
	
	// Static or random addr
	printf("MAC Address: ");
	if (change_class)
		printf("Randomized\n");
	else
		printf("Default\n");	

	// Static or random class
	printf("Device Class: ");
	if (change_class)
		printf("Randomized\n");
	else
		printf("Static (%s)\n", DEFAULT_CLASS);

	printf("Fogging started at [%s] on %i threads.\n", get_localtime(), numthreads);
	printf("Hit Ctrl+C to end.\n");	
	
	for( t = 0; t < numthreads; t++ )
	{
		// Thread ID number
		thread_data_array[t].thread_id = t;
		
		// Device number
		if (device == -1)
			thread_data_array[t].device = t;
		else
			thread_data_array[t].device = device;
			
		// Default information for all threads
		thread_data_array[t].change_addr = change_addr;
		thread_data_array[t].change_class = change_class;
		thread_data_array[t].static_name = static_name;
		thread_data_array[t].delay = delay;
		thread_data_array[t].loiter = loiter;
		
		// Start thread
		pthread_create(&threads[t], NULL, thread_spoof, (void *) &thread_data_array[t]);
			
		// Sleep for a second to stagger threads (needs experimentation)
		if (numthreads > 1)
			sleep (delay / numthreads);
	}
	
	// Install handler after threads started, terrible hack
	struct sigaction sig_handler;
	sig_handler.sa_handler = sig_catch;
	sigemptyset(&sig_handler.sa_mask);
	sig_handler.sa_flags = 0;
	sigaction(SIGINT, &sig_handler, NULL);
			    
    // Wait for Crtl+C
    pause();
    
    // Tell threads to wrap it up
	printf("\rTelling threads to shut down, this make take several seconds...\n");
	end_threads = 1;
	
	// Wait for threads to complete
	for ( t = 0; t < numthreads; t++ )
		pthread_join(threads[t], NULL);
			
	// Close up
	printf("All threads done.\n");
	exit(0);
}
